File: desktop/src/main/kotlin/com/reelvault/ui/screens/OpenCatalogDialog.kt
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 ReelVault Contributors

@file:OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)

package com.reelvault.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.DialogProperties
import com.reelvault.data.server.RecentCatalogs
import com.reelvault.ui.components.Tooltip
import com.reelvault.ui.theme.ReelVaultSpacing
import java.io.File

/**
 * Dialog shown when the client can't find a catalog to open — at startup
 * (after a fresh launch) or after the user picks File → Open.
 *
 * Lists recent catalogs, lets the user browse for an existing `.db` file,
 * or pick a directory in which a new catalog will be created.
 *
 * The dialog never closes itself; the caller must collapse it by setting
 * the binding to `false` when [onPick] returns.
 */
@Composable
fun OpenCatalogDialog(
    onDismiss: () -> Unit,
    onPick: (path: String) -> Unit,
    /** When `true`, the dialog has no Cancel button — only "Quit". Used for
     *  the startup-time required-pick scenario. */
    isStartup: Boolean = false,
    recents: RecentCatalogs = RecentCatalogs.Default,
) {
    val recentList = remember { recents.list() }

    AlertDialog(
        onDismissRequest = { if (!isStartup) onDismiss() },
        title = {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Icon(
                    imageVector = Icons.Default.LibraryBooks,
                    contentDescription = null,
                    modifier = Modifier.size(20.dp)
                )
                Spacer(modifier = Modifier.width(ReelVaultSpacing.Small))
                Text(if (isStartup) "Open a Catalog" else "Open Catalog")
            }
        },
        text = {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .heightIn(min = 200.dp, max = 480.dp)
            ) {
                Text(
                    text = "A catalog is the SQLite file ReelVault uses to remember " +
                        "your library — locations, tags, notes, thumbnails. Each catalog " +
                        "is independent.",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                Spacer(modifier = Modifier.height(ReelVaultSpacing.Medium))

                Row(horizontalArrangement = Arrangement.spacedBy(ReelVaultSpacing.Small)) {
                    Tooltip(text = "Pick an existing .db file from disk. ReelVault will mount it and load its library.") {
                        OutlinedButton(onClick = {
                            pickCatalogFile(forCreate = false)?.let { onPick(it) }
                        }) {
                            Icon(Icons.Default.FolderOpen, contentDescription = null, modifier = Modifier.size(16.dp))
                            Spacer(modifier = Modifier.width(4.dp))
                            Text("Open File…")
                        }
                    }
                    Tooltip(text = "Pick a folder and create a fresh catalog.db inside it. Use this for a brand-new library.") {
                        OutlinedButton(onClick = {
                            pickCatalogFile(forCreate = true)?.let { onPick(it) }
                        }) {
                            Icon(Icons.Default.CreateNewFolder, contentDescription = null, modifier = Modifier.size(16.dp))
                            Spacer(modifier = Modifier.width(4.dp))
                            Text("New Catalog…")
                        }
                    }
                }

                Spacer(modifier = Modifier.height(ReelVaultSpacing.Medium))
                Text(
                    text = "Recent",
                    style = MaterialTheme.typography.labelMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )

                if (recentList.isEmpty()) {
                    Text(
                        text = "No recent catalogs yet.",
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.outline,
                        modifier = Modifier.padding(vertical = ReelVaultSpacing.Small)
                    )
                } else {
                    LazyColumn(modifier = Modifier.fillMaxWidth()) {
                        items(recentList) { path ->
                            RecentRow(
                                path = path,
                                onOpen = { onPick(path) },
                                onForget = { recents.remove(path) }
                            )
                        }
                    }
                }
            }
        },
        confirmButton = {
            if (isStartup) {
                Tooltip(text = "Quit ReelVault without opening a catalog.") {
                    TextButton(onClick = {
                        // ExitApplication isn't available here; fall back to System.exit.
                        // The app harness uses a windowed `application { }` block so
                        // closing the JVM is fine.
                        System.exit(0)
                    }) {
                        Text("Quit")
                    }
                }
            } else {
                Tooltip(text = "Close this dialog without changing the currently open catalog.") {
                    Button(onClick = onDismiss) { Text("Cancel") }
                }
            }
        },
        properties = DialogProperties(usePlatformDefaultWidth = false),
        modifier = Modifier.width(560.dp)
    )
}

@Composable
private fun RecentRow(path: String, onOpen: () -> Unit, onForget: () -> Unit) {
    val file = remember(path) { File(path) }
    val display = remember(path) { file.nameWithoutExtension.ifEmpty { file.name } }
    val exists = remember(path) { file.exists() }
    val dirty = remember(path) { file.parentFile?.absolutePath ?: "" }

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(enabled = exists, onClick = onOpen)
            .padding(vertical = ReelVaultSpacing.XSmall, horizontal = ReelVaultSpacing.XSmall),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Icon(
            imageVector = if (exists) Icons.Default.LibraryBooks else Icons.Default.Error,
            contentDescription = null,
            modifier = Modifier.size(18.dp),
            tint = if (exists) MaterialTheme.colorScheme.primary else MaterialTheme.colorScheme.error
        )
        Spacer(modifier = Modifier.width(ReelVaultSpacing.Small))
        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = display,
                style = MaterialTheme.typography.bodyMedium,
                color = if (exists) MaterialTheme.colorScheme.onSurface
                        else MaterialTheme.colorScheme.onSurfaceVariant,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis
            )
            Text(
                text = if (exists) dirty else "$dirty (missing)",
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.outline,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis
            )
        }
        Tooltip(text = "Remove this catalog from the recent list. The file isn't deleted.") {
            IconButton(onClick = onForget, modifier = Modifier.size(28.dp)) {
                Icon(
                    imageVector = Icons.Default.Close,
                    contentDescription = "Forget",
                    modifier = Modifier.size(16.dp),
                    tint = MaterialTheme.colorScheme.outline
                )
            }
        }
    }
}

/**
 * Native open/save dialog for picking (or creating) a `.db` catalog file.
 * Returns the chosen absolute path, or `null` if the user cancelled.
 */
private fun pickCatalogFile(forCreate: Boolean): String? {
    val mode = if (forCreate) java.awt.FileDialog.SAVE else java.awt.FileDialog.LOAD
    val dialog = java.awt.FileDialog(null as java.awt.Frame?,
        if (forCreate) "Create New Catalog" else "Open Catalog", mode)
    if (forCreate) {
        dialog.file = "catalog.db"
    } else {
        // AWT's FileDialog filter API is limited; just accept anything.
    }
    dialog.isVisible = true
    val dir = dialog.directory ?: return null
    val file = dialog.file ?: return null
    return java.io.File(dir, file).absolutePath
}

File: macos/ReelVault/Views/OpenCatalogDialog.swift
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 ReelVault Contributors

import SwiftUI
import ReelVaultKit
import AppKit

/// Dialog shown when the daemon has no catalog mounted — at first launch
/// (after a clean install) or after the user picks File → Open. Lists recent
/// catalogs and provides "Open File…" / "New Catalog…" buttons.
///
/// The caller controls dismissal: when `onPick` is invoked, the parent should
/// set the binding to `false` once it has asked the daemon to mount the new
/// catalog.
struct OpenCatalogDialog: View {
    @Binding var isPresented: Bool
    let onPick: (_ path: String) -> Void
    /// When `true`, the only "back-out" action is "Quit ReelVault" — used at
    /// startup when there's no fallback catalog to keep using.
    let isStartup: Bool

    @ObservedObject private var recents = RecentCatalogs.shared
    @State private var hoveredPath: String? = nil

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            HStack {
                Image(systemName: "books.vertical.fill")
                Text(isStartup ? "Open a Catalog" : "Open Catalog")
                    .font(.headline)
                Spacer()
            }

            Text("A catalog is the SQLite file ReelVault uses to remember your library — locations, tags, notes, thumbnails. Each catalog is independent.")
                .font(.caption)
                .foregroundColor(.secondary)
                .fixedSize(horizontal: false, vertical: true)

            HStack(spacing: 8) {
                Button {
                    if let path = pickCatalogFile(create: false) {
                        onPick(path)
                    }
                } label: {
                    Label("Open File…", systemImage: "folder")
                }
                .help("Pick an existing .db file from disk. ReelVault will mount it and load its library.")

                Button {
                    if let path = pickCatalogFile(create: true) {
                        onPick(path)
                    }
                } label: {
                    Label("New Catalog…", systemImage: "doc.badge.plus")
                }
                .help("Pick a folder and create a fresh catalog.db inside it. Use this for a brand-new library.")
            }

            Divider()
            Text("Recent")
                .font(.caption)
                .foregroundColor(.secondary)

            let items = recents.list()
            if items.isEmpty {
                Text("No recent catalogs yet.")
                    .font(.caption)
                    .foregroundColor(.secondary)
                    .padding(.vertical, 4)
            } else {
                ScrollView {
                    VStack(alignment: .leading, spacing: 0) {
                        ForEach(items, id: \.self) { path in
                            recentRow(path: path)
                        }
                    }
                }
                .frame(maxHeight: 260)
            }

            HStack {
                Spacer()
                if isStartup {
                    Button("Quit") {
                        // `NSApp.terminate(nil)` is the "polite" path but it
                        // gets deferred while a modal sheet session is
                        // active — clicking Quit from inside the sheet
                        // makes the request never fire. We're at startup
                        // with no catalog mounted and nothing to save, so
                        // dismiss the sheet first and then hard-exit on the
                        // next runloop tick.
                        isPresented = false
                        DispatchQueue.main.async {
                            NSApp.terminate(nil)
                            // Belt-and-suspenders: if terminate() is still
                            // intercepted by something (SwiftUI scenes
                            // sometimes hold on), force exit shortly after.
                            DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
                                exit(0)
                            }
                        }
                    }
                    .keyboardShortcut("q", modifiers: [.command])
                    .help("Quit ReelVault without opening a catalog.")
                } else {
                    Button("Cancel") { isPresented = false }
                        .keyboardShortcut(.cancelAction)
                        .help("Close this dialog without changing the currently open catalog.")
                }
            }
        }
        .padding(20)
        .frame(width: 560)
    }

    @ViewBuilder
    private func recentRow(path: String) -> some View {
        let url = URL(fileURLWithPath: path)
        let exists = FileManager.default.fileExists(atPath: path)
        let display = url.deletingPathExtension().lastPathComponent
        let dir = url.deletingLastPathComponent().path

        HStack(spacing: 8) {
            Image(systemName: exists ? "books.vertical" : "exclamationmark.triangle.fill")
                .foregroundColor(exists ? .accentColor : .red)
            VStack(alignment: .leading, spacing: 2) {
                Text(display.isEmpty ? url.lastPathComponent : display)
                    .font(.system(size: 12))
                    .foregroundColor(exists ? .primary : .secondary)
                Text(exists ? dir : "\(dir) (missing)")
                    .font(.system(size: 10))
                    .foregroundColor(.secondary)
                    .lineLimit(1)
                    .truncationMode(.middle)
            }
            Spacer()
            Button {
                recents.remove(path)
            } label: {
                Image(systemName: "xmark.circle.fill")
                    .foregroundColor(.secondary)
            }
            .buttonStyle(.plain)
            .help("Remove this catalog from the recent list. The file isn't deleted.")
        }
        .padding(.horizontal, 6)
        .padding(.vertical, 4)
        .background(hoveredPath == path ? Color.accentColor.opacity(0.12) : Color.clear)
        .cornerRadius(4)
        .contentShape(Rectangle())
        .onHover { hovering in
            hoveredPath = hovering ? path : nil
        }
        .onTapGesture {
            if exists { onPick(path) }
        }
    }

    /// Native NSOpenPanel for picking (or NSSavePanel for creating) a `.db`
    /// catalog file. Returns the chosen absolute path or `nil` if cancelled.
    private func pickCatalogFile(create: Bool) -> String? {
        if create {
            let panel = NSSavePanel()
            panel.title = "Create New Catalog"
            panel.nameFieldStringValue = "catalog.db"
            panel.allowedContentTypes = []
            panel.canCreateDirectories = true
            return panel.runModal() == .OK ? panel.url?.path : nil
        } else {
            let panel = NSOpenPanel()
            panel.title = "Open Catalog"
            panel.canChooseFiles = true
            panel.canChooseDirectories = false
            panel.allowsMultipleSelection = false
            return panel.runModal() == .OK ? panel.url?.path : nil
        }
    }
}
